src/sql/src/statements/option_map.rs

```diff
@@ -12,52 +12,144 @@
+        );
+        map
+    }
+
+    pub fn into_map(self) -> HashMap<String, String> {
+        let mut map = self.options;
+        map.extend(
+            self.secrets
+                .into_iter()
+                .map(|(k, v)| (k, v.expose_secret().to_string())),
+        );
+        map
+    }
+
+    pub fn kv_pairs(&self) -> Vec<String> {
+        let mut result = Vec::with_capacity(self.options.len() + self.secrets.len());
+        for key in self.options.keys().sorted() {
+            if let Some(val) = self.options.get(key) {
+                result.push(format!("{key} = '{}'", val.escape_default()));
+            }
+        }
+        for key in self.secrets.keys().sorted() {
+            if let Some(val) = self.secrets.get(key) {
+                result.push(format!("{key} = {val:?}"));
+            }
+        }
+        result
     }
 }
 
 impl From<HashMap<String, String>> for OptionMap {
-    fn from(map: HashMap<String, String>) -> Self {
-        Self { map }
+    fn from(value: HashMap<String, String>) -> Self {
+        let mut result = OptionMap {
+            options: Default::default(),
+            secrets: Default::default(),
+        };
+        for (k, v) in value.into_iter() {
+            result.insert(k, v);
+        }
+        result
     }
 }
 
-impl AsRef<HashMap<String, String>> for OptionMap {
-    fn as_ref(&self) -> &HashMap<String, String> {
-        &self.map
+impl PartialEq for OptionMap {
+    fn eq(&self, other: &Self) -> bool {
+        if self.options.ne(&other.options) {
+            return false;
+        }
+
+        if self.secrets.len() != other.secrets.len() {
+            return false;
+        }
+
+        self.secrets.iter().all(|(key, value)| {
+            other
+                .secrets
+                .get(key)
+                .map_or(false, |v| value.expose_secret() == v.expose_secret())
+        })
     }
 }
 
-impl Borrow<HashMap<String, String>> for OptionMap {
-    fn borrow(&self) -> &HashMap<String, String> {
-        &self.map
+impl Eq for OptionMap {}
+
+impl Visit for OptionMap {
+    fn visit<V: Visitor>(&self, visitor: &mut V) -> ControlFlow<V::Break> {
+        for (k, v) in &self.options {
+            k.visit(visitor)?;
+            v.visit(visitor)?;
+        }
+        for (k, v) in &self.secrets {
+            k.visit(visitor)?;
+            v.expose_secret().visit(visitor)?;
+        }
+        ControlFlow::Continue(())
     }
 }
 
-impl FromIterator<(String, String)> for OptionMap {
-    fn from_iter<I: IntoIterator<Item = (String, String)>>(iter: I) -> Self {
-        Self {
-            map: iter.into_iter().collect(),
+impl VisitMut for OptionMap {
+    fn visit<V: VisitorMut>(&mut self, visitor: &mut V) -> ControlFlow<V::Break> {
+        for (_, v) in self.options.iter_mut() {
+            v.visit(visitor)?;
+        }
+        for (_, v) in self.secrets.iter_mut() {
+            let mut v = v.expose_secret().to_string();
```

Comment:
Extra clone. IIRC new nightly version of `secrecy` has an `expose_secret_mut` method for ease this case.

Reply:
Resolved.